hash.c

```diff
@@ -4377,18 +4377,29 @@ rb_hash_compare_by_id(VALUE hash)
     if (hash_iterating_p(hash)) {
         rb_raise(rb_eRuntimeError, "compare_by_identity during iteration");
     }
-    ar_force_convert_table(hash, __FILE__, __LINE__);
-    HASH_ASSERT(RHASH_ST_TABLE_P(hash));
 
-    tmp = hash_alloc(0);
-    hash_st_table_init(tmp, &identhash, RHASH_SIZE(hash));
-    identtable = RHASH_ST_TABLE(tmp);
+    if (RHASH_TABLE_EMPTY_P(hash)) {
+        // Fast path: There's nothing to rehash, so we don't need a `tmp` table.
+        // We're most likely an AR table, so this will need an allocation.
+        ar_force_convert_table(hash, __FILE__, __LINE__);
+        HASH_ASSERT(RHASH_ST_TABLE_P(hash));
 
-    rb_hash_foreach(hash, rb_hash_rehash_i, (VALUE)tmp);
+        RHASH_ST_TABLE(hash)->type = &identhash;
+    } else {
+        // Slow path: Need to rehash the members of `self` into a new
+        // `tmp` table using the new `identhash` compare/hash functions.
+        tmp = hash_alloc(0);
+        hash_st_table_init(tmp, &identhash, RHASH_SIZE(hash));
```

Comment:
Could this just re-use `rb_ident_hash_new_with_size`?

 ```suggestion
        tmp = rb_ident_hash_new_with_size( RHASH_SIZE(hash));
        identtable = RHASH_ST_TABLE(tmp);
```

What's the difference between `hash_alloc(0)` and `rb_hash_new`?

It calls `hash_alloc(rb_cHash)` instead, but I don't understand the significance of passing `0` (null?) in this case.

Reply:
We want to have a `klass` of 0 because we want this hash to be hidden (i.e. it does not get outputted in the `ObjectSpace` traversals).

Reply:
Is that a common convention? Seems a bit magical. Is it worth a comment?

Or perhaps being more explicit with `rb_obj_hide(rb_ident_hash_new_with_size(RHASH_SIZE(hash)))`?

Reply:
Maybe we can have a `rb_hash_hidden_new` to mirror `rb_ary_hidden_new`.